airflow/models/taskinstance.py

```diff
@@ -2580,6 +2580,31 @@ def get_num_running_task_instances(self, session: Session, same_dagrun=False) ->
             )
         return num_running_task_instances_query.scalar()
 
+    @provide_session
+    def get_valid_map_index(self, session: Session, limit: int) -> (set[int]):
+        """Return running map index of task group from the DB."""
```

Comment:
Can logic in this function and TaskGroup be unified?

Reply:
You mean the code to generate concurrency_map in scheduler_job_runer.py, right?
This is a good idea. But the two places have some differences.

Code here is for task_concurrency_dep.py, which checks for single TI before it be queued.
And the code for concurrency_map is for all TIs after being queued.
So their logic is not very similar.